### PR TITLE
Alternate solution for timeout blur

### PR DESCRIPTION
**Pull request by bartpeeters**
Follow up from https://github.com/gpbl/react-day-picker/issues/579#issuecomment-353798385, should replace #525.

---

Issue:
1. Open Firefox (its not an issue in chrome) https://codesandbox.io/s/XDAE3x0W8
2. Click on day picker input (it shows calendar)
3. Click on navigation bar element (next, prev or nav title elements)
4. Main day picker input loses own focus and calendar stays open
5. Now when you click anywhere outside of calendar. Calendar is still open but is should close.

The fix using a setTimeout onBlur to call this.input.focus() caused new issues: 
- A flicker in Chrome everytime you click inside the daypicker
- When using a custom captionElement containing a <select> it would not open because the timeout focus cancelled it

This alternate solution is to remove the timeout and instead add an onBlur on the overlay doing the same thing as the onBlur of the input.

### CODE REVIEW

**Comment by gpbl** [comment]
Thanks @bartpeeters for your PR. I want to try it first, sure we have to remove that timeout thing since it doesn't well. 

I don't understand why the new `span` element should fire the `onBlur` event [here](https://github.com/gpbl/react-day-picker/pull/598/files#diff-e2a925e665d836a09575f79095191746R415). It shouldn't ever get the focus. Have you find the reason behind it?

**Comment by bartpeeters** [comment]
@gpbl tabIndex on the daypicker ([here](https://github.com/gpbl/react-day-picker/blob/8c69e20419bfc3a601de7ee267c2436707f3ed62/src/DayPicker.js#L553)) makes it so the span is focusable.

File: src/DayPickerInput.js [hunk 1]
```diff
@@ -417,22 +412,24 @@ export default class DayPickerInput extends React.Component {
...
-        classNames={classNames}
-        month={this.state.month}
-        selectedDay={selectedDay}
-        input={this.input}
-      >
-        <DayPicker
-          ref={el => (this.daypicker = el)}
-          onTodayButtonClick={onTodayButtonClick}
-          {...dayPickerProps}
+      <span onBlur={this.handleInputBlur}>
```
Comment (gpbl) [inline src/DayPickerInput.js]: My suspect is that this line does nothing. We should add an unit test to it.

File: test/daypickerinput/events.js [hunk 2]
```diff
@@ -52,18 +52,6 @@ describe('DayPickerInput', () => {
         wrapper.find('input').simulate('blur');
         expect(onBlur).toHaveBeenCalledTimes(1);
       });
-      it('should focus the input if blur after clicking the overlay', done => {
```
Comment (gpbl) [inline test/daypickerinput/events.js]: This test must stay here :) Why does your change make it fail?

**Comment by bartpeeters** [comment]
@gpbl Thanks for the feedback.
I implemented another solution, using the onFocus event of the overlay to always make sure the DaypickerInput input field is focused. 

But do we want this? What if we have a custom overlay that includes an input? Now this input is unusable because it can't be focused. We can check if the nodeName of the focus event target is an input, but I don't really know if that is the way to go.

I also replaced the clickInside with the timeout by using a ref on the Overlay and checking if the relatedTarget of blur event is inside the overlay. I'm not sure that this is better than using a timeout.

About the bubbling of the onFocus and onBlur on the span which contains the overlay, the React Synthetic onFocus and onBlur events do bubble, unlike the normal blur and focus events.

**Comment by gpbl** [comment]
@bartpeeters thanks for the updates. I've not checked your changes yet, however I'm proposing a different solution:

- reset this line to the original behavior, which was working well except on Firefox:

```diff
if (this.clickedInside) { 
-    this.blurTimeout = setTimeout(() => this.input.focus(), 0); 
+   this.input.focus(); 
}
```

- on Firefox, we just disable this to resolve https://github.com/gpbl/react-day-picker/issues/579:

```diff
- if (this.clickedInside) { 
+ if (!isFirefox && this.clickedInside) { 
```

- for the case when the calendar has other focusable elements inside, such as selects or input fields (as in https://github.com/gpbl/react-day-picker/issues/599), we allow the implementer to disable this focus with a new DayPickerInput prop: `keepFocus` that can be set to `false`.

**Comment by bartpeeters** [comment]
@gpbl I like the idea of a new prop `keepFocus`.

The solution you propose has the same effect as these changes except they dont use `isFirefox`, so maybe you can take a look at the code. I'm also up to add the `keepFocus` prop and some tests if you approve of these changes.

**Comment by mattfysh** [comment]
@gpbl how can I help push this forward? would really appreciate being able to set `keepFocus` to `false`

**Comment by bartpeeters** [comment]
@mattfysh I just added the `keepFocus` prop and added api documentation for it. I think now we just need to wait till @gpbl has some time to review these changes.

File: docs/src/pages/api/DayPickerInput.js [hunk 3]
```diff
@@ -246,6 +247,16 @@ function MyDayPickerInput(props) {
       <p>
         The value of the <code>input</code> field.
       </p>
+      <h3>
+        <Anchor id="keepFocus" />
+        keeFocus <code>boolean = true</code>
```
Comment (ghaynes1292) [inline docs/src/pages/api/DayPickerInput.js]: keepFocus

Comment (bartpeeters) [reply docs/src/pages/api/DayPickerInput.js]: Thanks!

**Comment by rdiwelm** [comment]
Hey guys! How soon can we expect a new version release with the new `keepFocus` prop?

**Comment by gpbl** [comment]
@rdiwelm soon ! I have been pretty busy the last month, hoping to catch up all this weekend 🙏🏽 sorry for the lack of communication.

**Comment by roginfarrer** [comment]
I also have a feature depending on this release!

**Comment by rdiwelm** [comment]
Hey guys! Sorry to be asking again. Is there any ETA on this? We're kinda blocked too. :/

**Comment by kradical** [comment]
@roginfarrer @rdiwelm I was blocked by this too so we forked this dependency until this change gets merged. I understand that is not always a viable option.

**Comment by gpbl** [comment]
Published as [v7.1.0](https://www.npmjs.com/package/react-day-picker).

**Comment by rdiwelm** [comment]
Awesome! Thanks a lot, guys! :)

**Pull request merged**